lib/rex/proto/ldap/auth.rb

```diff
@@ -0,0 +1,353 @@
+        # @param dnsname   [String] DNS Name value used in NTLM
+        # @param dnsdomain [String] DNS Domain value used in NTLM
+        def initialize(challenge = Rex::Text.rand_text_alphanumeric(16), domain  = 'DOMAIN',
+                       server = 'SERVER', dnsname = 'server', dnsdomain = 'example.com')
+          @domain = domain
+          @server = server
+          @dnsname = dnsname
+          @dnsdomain = dnsdomain
+          @challenge = [challenge].pack('H*')
+        end
+
+        #
+        # Process the incoming LDAP login requests from clients
+        #
+        # @param user_login [OpennStruct] User login information
+        #
+        # @return auth_info [Hash] Processed authentication information
+        def process_login_request(user_login)
+          auth_info = {}
+
+          if user_login.name.empty? && user_login.authentication.empty? # Anonymous
+            auth_info = handle_anonymous_request(user_login, auth_info)
+          elsif !user_login.name.empty? # Simple
+            auth_info = handle_simple_request(user_login, auth_info)
+          elsif user_login.authentication[0] == 'GSS-SPNEGO' # SASL especially SPNEGO
+            auth_info = handle_sasl_request(user_login, auth_info)
+          else
+            auth_info[:result_code] = Net::LDAP::ResultCodeUnwillingToPerform
+          end
+
+          auth_info
+        end
+
+        #
+        # Handle Anonymous authentication requests
+        #
+        # @param user_login [OpennStruct] User login information
+        # @param auth_info [Hash] Processed authentication information
+        # 
+        # @return auth_info [Hash] Processed authentication information
+        def handle_anonymous_request(user_login, auth_info = {})
+          if user_login.name.empty? && user_login.authentication.empty?
+            auth_info[:user] = user_login.name
+            auth_info[:pass] = user_login.authentication
+            auth_info[:domain] = nil
+            auth_info[:result_code] = Net::LDAP::ResultCodeSuccess
+            auth_info[:auth_type] = 'Anonymous'
+
+            auth_info
+          end
+        end
+
+        #
+        # Handle Simple authentication requests
+        #
+        # @param user_login [OpennStruct] User login information
+        # @param auth_info [Hash] Processed authentication information
+        # 
+        # @return auth_info [Hash] Processed authentication information
+        def handle_simple_request(user_login, auth_info = {})
+          domains = []
+          names = []
+          if !user_login.name.empty?
+            if user_login.name =~ /@/
+              pub_info = user_login.name.split('@')
+              if pub_info.length <= 2
+                auth_info[:user], auth_info[:domain] = pub_info
+              else
+                auth_info[:result_code] = Net::LDAP::ResultCodeInvalidCredentials
+                auth_info[:result_message] = "LDAP Login Attempt => From:#{auth_info[:name]} DN:#{user_login.name}"
```

Comment:
`auth_info[:name]` does not look to be set when this code is reached. These are edge cases so adding `rspec` should further detail this.